🐈‍⬛ 1단계 - 로그인 구현 리뷰 요청

parkjun5 opened the pull request:
안녕하세요!

로그인 구현을 마치고 리뷰를 요청드립니다.

생각보다 기능을 이것 저것 넣어야 하다보니 아직 테스트도 다 추가해두진 않았지만,
시간이 너무 지체되는 것 같아 우선 리뷰 신청합니다!

이후 열심히 테스트 추가하겠슴다..ㅎㅎ

리뷰 잘 부탁드립니다.

parkjun5 committed:
😺 Step1 1단계 채리픽 - 예약 기능 리뷰 요청 (#53)

* feat: Define Reservation Domain And ReservationTime Domain

* feat: Add JDBC repositories for Create, Read, Delete operations

* feat: Add common support settings to the project

- Add index controller to support pages.
- ReservationDateChecker to validate dates.
- Add properties datasource setting.

* feat: Implement Reservation Application Layer

- Add ReservationCanceler to cancel Reservation.
- Add ReservationFinder to find reservations.
- Add ReservationRecorder to record new reservations.

* feat: Implement Reservation Presentation Layer

* feat: Implement ReservationTime Application Layer

- Add ReservationTimeDeleter to delete ReservationTime.
- Add ReservationTimeFinder to find reservationsTime.
- Add ReservationTimeSaver to save new reservationsTime.

* feat: Implement ReservationTime Presentation Layer

* test: Update tests for reservation changes and add new feature for reservation time

* test: Add more application tests

- Add tests for reservation applications.
- Add tests for reservation time applications.

---------

Co-authored-by: MZC01-PARKJUN5 <parkjun5@mz.co.kr>

(cherry picked from commit 43cdfdb53299f0b668e5224217cf16143c93535b)

parkjun5 committed:
😺 2단계 - 예외 처리 및 리팩터링 리뷰 요청 채리픽 (#64)

* fix: theme 추가로 인한 링크 추가 및 기존 링크 수정

* feat: theme 추가로 인한 html 및 js 파일 생성

* refactor: 리뷰 내용 적용

- 패키지 명 변경 time -> reservationtime
- delete 메서드 명 변경 delete -> deleteById
- usingRecursiveComparison 적용

* refactor: change reservation and add Theme

* refactor: add theme support to reservation application, repository, and UI layers

* feat: Implement Theme Application Layer

- Add ThemeDeleter to delete theme.
- Add ThemeFinder to find themes.
- Add ThemeSaver to save new theme.

* feat: Define Theme Domain

* feat: Implement Theme Presentation Layer

* test: Add more Test for Theme Domain

* feat: Add GlobalExceptionHandler to handle application errors and define specific exceptions

* etc: add theme page and change admin pages uri

* test: Add theme application tests

* polishing

* refactor: 리뷰 내용 적용

- 예약 시간, 테마 삭제시 예약이 존재하는 지 확인 후 삭제하도록 변경(예약이 있으면 삭제 불가하도록 Table 변경)
- GlobalExceptionHandler가 좀더 명확하게 처리하도록 변경
    - 해당 변경점을 위한 테스트 추가

---------

Co-authored-by: boorownie <probitanima11@gmail.com>

Commit:
polishing

parkjun5 committed:
feat: develop member and member-role domain

 - 로그인을 위한 맴버, 댐버 롤 도메인 추가

parkjun5 committed:
feat: develop login domain with jwt token

 - jwt 토큰을 사용하는 로그인 도메인 개발

parkjun5 committed:
etc: 1단계 진행을 위한 html, js 주석 해제

 - IndexController 경로 변경

parkjun5 committed:
feat: 예약가능한 테마의 시간 찾기 API 개발

parkjun5 committed:
feat: change http-status not just 200

 - create -> created(201)
 - delete -> no_content(204)

parkjun5 committed:
feat: chanage JdbcTemplate to NamedParameterJdbcTemplate

 - 복잡한 쿼리가 늘면서 파리미터를 명시적으로 확인하기 좋은 파라미터 템플릿으로 변경

parkjun5 committed:
polishing

parkjun5 committed:
polishing

NewWisdom commented on src/main/java/roomescape/apply/auth/ui/dto/LoginRequest.java:
```diff
@@ -0,0 +1,28 @@
+
+import org.springframework.util.StringUtils;
+
+public record LoginRequest(
```
record 사용 👍

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -0,0 +1,70 @@
+
+    public String generateToken(LoginResponse loginResponse) {
+        LocalDateTime now = LocalDateTime.now();
+        LocalDateTime expiryDate = now.plusMinutes(EXPIRATION_TIME);
```
EXPIRATION_TIME 보다 Minutes(분)을 드러내는 네이밍은 어떨까요~?

parkjun5 replied:
시간이라는 표현이 좀 애매하네요. 이 부분에 많은 모호한 표현이 많으니 아예 분리해서 토큰의 시간을 관리해주는 클래스를 만들어 보았습니다.

TokenValidityPeriodCreator 에서 이제 토큰의 만료 시간관리를 책임져주려고 합니다.

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -0,0 +1,70 @@
+                .subject(loginResponse.email())
+                .claim("name", loginResponse.name())
+                .claim("role", loginResponse.memberRoleNames())
+                .issuedAt(Timestamp.valueOf(now))
```
LocalDate 가 아닌 Timestamp 를 사용해주셨네요.   
Timestamp 는 `java.sql` 패키지에 속해있는데요, 해당 클래스의 역할을 다시 살펴보고 적절한 사용인지 고민해보면 좋을 것 같네요 :)

parkjun5 replied:
이부분은 깊게 생각하지 않고 Timestamp으로 변경해서 사용하고 하였는데, 확인해보니 java.util의 Date를 받고 있었네요.

해당 변수를 변경하도록 바꿔보았습니다.

그리고 LocalDateTime(만료시간 추가) -> Instance -> Timestamp or Date 로 변환하고 있었는데,  Instance(만료시간 추가)에서 바로 Date 가도 될 것 같네요

NewWisdom replied:
issuedAt 에 LocalDate 타입을 넣을 수 있을 것 같네요, 
Date 와 LocalDate 는 어떤 차이가 있을까요~?

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -0,0 +1,70 @@
+                .compact();
+    }
+
+    public void validateToken(String token) {
+        try {
+            parseJwt(token);
+        } catch (ExpiredJwtException e) {
+            throw new IllegalTokenException(NOT_VALID_TOKEN_ERROR, e);
+        } catch (RuntimeException e) {
+            throw new IllegalTokenException(e);
+        }
```
상세한 예외 처리 👍 
ExpiredJwtException, RuntimeException 로 나뉘어서 예외 처리를 해주셨는데 결국 동일한 메시지의 IllegalTokenException 로 처리되겠네요!
ExpiredJwtException 는 만료된 토큰의 의미를 담은 커스텀 예외를 지정해도 좋을 것 같아요!

parkjun5 replied:
두개의 예외를 별도로 처리해주는 듯 하지만 실제로는 별 다르지가 않았네요.

별도의 메시지를 넣어줘야 겠네요.

이런 경우 ExpiredJwtException에러는 예상된 에러기 때문에, Throwable을 넘길 필요 없을 수 도 있겠네요

NewWisdom commented on src/main/java/roomescape/apply/auth/application/exception/IllegalTokenException.java:
```diff
@@ -0,0 +1,18 @@
+    public IllegalTokenException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}
```
No newline at end of file 파일 끝에 개행이 누락되었네요!
[해당 글](https://hstory0208.tistory.com/entry/IntelliJ-%ED%8C%8C%EC%9D%BC-%EB%81%9D%EC%97%90-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EA%B0%9C%ED%96%89-%EC%B6%94%EA%B0%80%ED%95%98%EA%B8%B0-No-newline-at-end-of-file)을 참고해서 자동으로 개행되도록 설정해도 좋을 것 같아요 :)

NewWisdom commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,64 @@
+CREATE TABLE reservation_time
```
data.sql 에 DDL, DML 이 모두 정의되어 있는데, schema.sql 도 이용해 DDL, DML 역할에 맞게 sql 문을 분리하는 것은 어떨까요~?

parkjun5 replied:
현재는 간단하게 data.sql로 진행하고 있었는데, 점점 복잡해지니 분리해주면 좋겠네요

NewWisdom replied:
👍 
참고로 더미 데이터를 생성하는 방법은 sql 문 말고도 `@ApplicationRunner` 를 활용하는 방법도 있으니 괜찮으면 시도해보아도 좋겠네요!

parkjun5 replied:
그 방법을 사용하면 DB에 직접 Insert가 아니라 스프링 어플리케이션을 통해 데이터가 들어가면서 값 삽입과 동시에 정상 케이스의 테스트가 될 수 도 있겠네요!

NewWisdom commented on src/main/java/roomescape/apply/reservation/application/ReservationRecorder.java:
```diff
@@ -0,0 +1,43 @@
+import roomescape.apply.theme.domain.Theme;
+
+@Service
+public class ReservationRecorder {
```
현재 application 레이어에 있는 서비스들을 보니, 영속화 시키는 행위에 많은 용어가 쓰이고 있네요 (adder, saver, recorder...)
각각 네이밍이 나온 이유가 궁금해요!

parkjun5 replied:
처음에는 도메인에 맞춰서 서비스  이름도 변하는게 자연스럽다고 생각했어요.

사람은 새로운 사람이 추가된 것이라 Adder이고 예약 시간은 새로 저장된 것이라 Saver 처럼 작성하였습니다.

이러면 메서드 이름을 지을때도 자연스럽고 역할도 클래스 이름만을 보고 좀더 이해하기 좋지 않을까 하였습니다 ㅎㅎㅎ...

그런데 작성하다보니 어떤 것이 Adder인지 Saver인지 기준이 참 애매하네요.
저 혼자 작성할때에는 쭈욱 작성하면 되지만, 팀에서 일한다면 이런 네이밍 컨벤션 규칙을 잡는게 쉽지 않겠네요.

NewWisdom replied:
좋은 의견이네요 👍 
개인적으로는 현재 미션에서는 예약, 예약 시간 등에 대해 단순하게 `생성`이라는 비즈니스가 있는 것 같은데 이에 대해 다양한 용어가 사용되면 오히려 혼란이 생길 수 있다고 생각했어요! 
말씀 주신 것처럼 타인이 읽었을 때 모호함도 생길 수 있겠네요 :)

NewWisdom commented on src/main/java/roomescape/support/checker/ReservationTimeRequestChecker.java:
```diff
@@ -0,0 +1,32 @@
+
+import org.springframework.util.StringUtils;
+
+public class ReservationTimeRequestChecker {
```
요청값에 대한 검증을 아주 꼼꼼히 해주셨네요 👍 
추후에 Spring Validation 을 통해 검증할 수도 있겠네요~!

parkjun5 replied:
Spring Validation 이나 Custom Annotaion으로 작업해보려고 합니다!

NewWisdom commented on src/main/java/roomescape/support/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,78 @@
+        logger.error("Exception caught: ", ex);
+        return new ResponseEntity<>("An error occurred", HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+    public record ExceptionResponse(
+            int status,
+            String message
+
+    ) {
```
이미 HTTP 응답 스펙으로 상태코드를 주고 있는데 수동으로 응답값에 상태코드를 넣은 이유는 무엇인가요~?

parkjun5 replied:
http 응답스펙으로도 상태코드를 주지만, body로도 넘겨주면 HTTP 상태 코드 이외의 커스텀 응답 코드를 넘겨줘서 관리하기 좋다고 생각합니다. 이러면 좀더 세밀한 예외 혹은 응답 처리가 가능하다고 생각해요.

NewWisdom replied:
여기서 사용하고 있는 `status` 가 HTTP 상태코드와 동일한 값을 쓰고 있어 중복해서 응답하고 있다고 생각했어요!
세준님 말씀을 들어보니 상태코드와 별개로 정의한 에러코드를 의도하신 것 같네요 :) 
HTTP 상태 코드와 다른 값의 에러코드로 정의해도 좋겠네요!

parkjun5 replied:
다음 미션에서 상태코드를 좀더 세분화해봐야겠네요!

NewWisdom commented on src/main/java/roomescape/apply/member/ui/MemberController.java:
```diff
@@ -0,0 +1,36 @@
+
+    @PostMapping
+    public ResponseEntity<MemberResponse> addMember(@RequestBody MemberRequest request) {
+        return ResponseEntity.status(HttpStatus.CREATED)
+                             .body(memberAdder.addNewMember(request));
```
```suggestion
        return ResponseEntity.ok(memberAdder.addNewMember(request));
```
위 처럼 사용할 수 있겠네요 ㅎㅎ

parkjun5 replied:
추천해주신 해당 코드로는 200으로 떨어지지 않을까요? 

Test에서 201코드를 확인하고 있으며, 클라이언트 코드에서도 201을 확인하고 있어서 ok로 하면 다르게 작동하지 않을 까 합니다.

NewWisdom replied:
아하 제가 예시 코드의 메서드를 잘못 적은 것 같은데 의도는 ResponseEntity 가 제공하는 정적 메서드를 사용하는 것이었어요!
```
	public static BodyBuilder created(URI location) {
		return status(HttpStatus.CREATED).location(location);
	}
```
URI까지 포함한다면 HTTP 201 스펙을 표현하는데 더 좋겠네요!

Commit:
test: fix old test

Commit:
test: Add Client Test For Present Layer

 - test new pages response
 - test member domain
 - test login domain
 - test find Available reservation time

Commit:
test: Add JwtTokenManagerTest

 - change method name generateToken to generateTokenByLoginResponse

Commit:
polishing

NewWisdom commented on src/main/java/roomescape/apply/auth/application/PasswordHasher.java:
```diff
@@ -0,0 +1,35 @@
+import java.security.NoSuchAlgorithmException;
+
+
+@Service
+public class PasswordHasher {
```
해당 클래스를 스프링 빈으로 관리하는 이유가 있나요?? 빈으로 관리하는 것의 장점은 무엇일까요~?

parkjun5 replied:
테스트할때에는 MockPasswordHasher로 처리하려고 스프링 빈으로 만들었습니다.

맴버 생성 테스트와 같은 테스트들에서는 암호화가 잘되는지는 관심사가 아니기 떄문에 이런 부분은 빈으로 등록하고 나중에 Mocking하는게 편하다고 생각해서 빈으로 등록하였습니다.

NewWisdom commented on src/main/java/roomescape/apply/auth/application/LoginManager.java:
```diff
@@ -0,0 +1,34 @@
+
+    public void addTokenToCookie(LoginResponse loginResponse, HttpServletResponse servletResponse) {
+        String token = jwtTokenManager.generateTokenByLoginResponse(loginResponse);
+        Cookie cookie = new Cookie(ServletRequestTokenFinder.COOKIE_NAME, token);
+        cookie.setHttpOnly(true);
+        cookie.setPath("/");
```
cookie 의 유효기간 값을 설정해보아도 좋겠네요!

parkjun5 replied:
토큰에 만료기간보다 1초더 줘서 토큰과 쿠키의 만료기간이 비슷하게 작동하면 좋읗 것 같네요!!

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -0,0 +1,68 @@
+
+        return Jwts.builder()
+                .subject(loginResponse.email())
+                .claim("name", loginResponse.name())
+                .claim("role", loginResponse.memberRoleNames())
```
name 과 role 은 중요한 값으로 보이는데 상수로 관리해도 좋을 것 같네요!

parkjun5 replied:
상수로 NAME_CLAIM, ROLE_CLAIM 이름이 좋아보여요!

NewWisdom commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,64 @@
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE member_role
```
member 테이블에 role을 두지 않고 따로 member_role 테이블을 두신 이유가 궁금해요!

parkjun5 replied:
한명의 사용자에게 여러 권한을 주기 위해서 입니다.  

사용자들에게 하나의 권한 이상이 필요하거나, 권한을 세분화해서 서비스별 나눠줘야 하는 경우가 발생한다고 생각합니다.

그래서 맴버와 권한이 1대N 관계가 되기 떄문에 별도의 테이블을 두었습니다.

NewWisdom commented on src/main/java/roomescape/apply/auth/application/JwtTokenManager.java:
```diff
@@ -0,0 +1,68 @@
+    private static final long EXPIRATION_TIME = 60;
+    private final SecretKey secretKey;
+
+    public JwtTokenManager(@Value("${jwt.secret}") String secretKey) {
```
`@Value("${jwt.secret}")` 를 통해 secret 을 가져오도록 잘 구현하셨네요 👍 
혹 여유가 되신다면 `@ConfigurationProperties` 에 대해 학습해보아도 좋을 것 같아요 :)

NewWisdom commented on src/test/java/roomescape/apply/auth/application/JwtTokenManagerTest.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    @DisplayName("토큰을 검증 할 수 있다.")
```
한 테스트 검증문에 유효한 토큰 검증, 유효하지 않은 토큰 케이스 모두 확인하고 있는데, 
한 테스트에 한가지 케이스를 확인하도록 명확히 분리하는 것은 어떠신가요~?

parkjun5 replied:
두가지 역할을 하는 테스트는 분리해주면 깔끔하네요!

NewWisdom reviewed (changes requested):
안녕하세요 세준님~! 
요구사항에 맞게 미션 잘 진행해주셨네요 👍 
많이 고민하신 흔적들이 보여서 재미있게 보았습니다!
몇가지 코멘트 남겨놓았는데 한번 확인해보시고 세준님의 생각대로 반영해주셔도 좋을 것 같습니다 :)

Commit:
test: Add More Test for member and member role application layers

Commit:
refactor: role name과 value가 혼동되는 문제를 명시적으로 메서드이름으로 표현

Commit:
refactor: 리뷰 적용

Commit:
refactor: 토큰 예외 세분화

Commit:
refactor: data.sql 분리

 - schema.sql과 data.sql으로 분리하여 데이터 조작 언어와 데이터 정의 언어를 별도로 관리

Commit:
refactor: 토큰 검증 테스트 분리

 - 잘못된 토큰 검증과 정상 토큰 검증 테스트 분리

NewWisdom commented on src/test/java/roomescape/FirstMissionStepTest.java:
```diff
@@ -0,0 +1,178 @@
+
+import static org.hamcrest.Matchers.is;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
```
테스트 잘 작성해주셨네요 👍 
실제 애플리케이션과 동일한 포트를 사용하고 있어 애플리케이션이 띄워져 있는 상태에서 테스트를 실행하면 실패하게 되네요!
서로 다른 포트를 사용하게 하는 것은 어떤가요~?

parkjun5 replied:
테스트는 별도로 포트를 두고, 포트를 test.resources의 폴더 아래 application.yml에 정의해서 컨트롤하면 좋읗 것 같아요!

NewWisdom commented on src/main/java/roomescape/apply/member/application/MemberAdder.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public MemberResponse addNewMember(MemberRequest request) {
+        duplicateChecker.checkIsDuplicateEmail(request);
```
현재 유즈 케이스 별로 서비스를 나누어 구현해주신 것 같은데요, 이렇게 구현하신 이유와 느끼신 장점 혹은 단점이 있을까요~?

parkjun5 replied:
유즈 케이스별로 서비스를 나눈 이유는 각 서비스들이 책임을 작게 가지게 해서 단순하게 구현하기 위해서입니다.

복잡하게 개발되어있는 것보다는 단순하게 개발되어 있는 것이 나중에 활용하기도 좋고 개발하기도 편하다고 생각해요.

장점은 빠르게 개발하고 유지보수 그리고 해당 클래스를 다른 클래스에서 사용하기 엄청 편합니다!
테스트 작성할 때에도 해당 기능만 테스트하면 되서 직관적이고요. 단순하게 만드는 것이 최고 장점이라고 생각해요.

단점은 아무래도 클래스가 많이 쪼개지다보니 코드가 많아지는 느낌도 있고, 프로젝트의 구조를 잘 나누지 않으면
아무래도 클래스들이 많다보니까 찾기 어렵더라구요 ㅎㅎ;
또 기능을 확장해가면서 어려움이 있는 것 같아요.
기존 Finder라는 클래스에 새로 기능을 추가할 지 아니면 별도의 Finder를 만들어야 할지 고민되는 부분이 생기더라구요.

NewWisdom replied:
👍 
현재 미션 정도의 요구사항에서는 memberService, ReservationService 등 서비스를 단순화 할 수 있지 않을까라는 생각을 했던 것 같아요! 현재 구조에서는 요구사항에 대해 로직이 여러 클래스들에 퍼져있어서 의존하고 있는 클래스들이 많아지고, 오히려 복잡도가 높아지는 부분이 우려되었던 것 같아요, 말씀해주신 대로 클래스를 구분하는 기준도 모호할 수 있구요 ㅎㅎ 사실 구현에 정답은 없으니 제 의견은 참고만 해주셔도 좋을 것 같아요 :)
고민하신 부분 잘 정리해서 남겨주셔서 감사합니다! 미션 진행하시면서 구조에 대해 깊게 고민하시는 모습이 인상적이네요 👍

NewWisdom commented on src/main/java/roomescape/apply/reservation/domain/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,23 @@
+
+    List<Reservation> findAll();
+
+    Optional<Long> checkIdExists(long id);
+
+    void deleteById(long id);
+
+    Optional<Long> findIdByTimeIdAndThemeId(long timeId, long themeId);
```
중복 여부를 체크하는 로직에 대해 `findId~`, `checkId~` 메서드명을 사용하기도 하고 count 를 조회해서 서비스 레이어에서 0인지 여부를 체크하기도 하는 것 같은데 전반적으로 통일성 있게 중복 여부를 체크하는 로직을 정리해보는 것은 어떨까요~?

parkjun5 replied:
한번 정리를 해야겠네요!! 중복 여부에서 중복된 횟수까지 알 필요는 없으니 findId로 가는게 깔끔할 것 같네요

NewWisdom commented on src/main/java/roomescape/support/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);
+
+    @ExceptionHandler({TokenNotFoundException.class, IllegalTokenException.class})
+    public ResponseEntity<ExceptionResponse> handleTokenException(IllegalArgumentException e) {
+        return ResponseEntity.badRequest().body(new ExceptionResponse(400, e.getMessage()));
+    }
```
인증 오류와 관련되어서 더 의미있는 HTTP 상태코드는 없을까요~?

parkjun5 replied:
401 을 보내주는 게 올바르겠네요. 아무래도 재로그인을 요청해야하기도 하니까요

NewWisdom reviewed (approved):
안녕하세요 세준님! 
피드백 빠르게 잘 반영해주셨네요 👍 
세준님의 생각이 궁금한 부분 코멘트 남겨두었는데 시간 되실 때 답변해주시면 감사하겠습니다 !
추가로 남긴 피드백은 다음 미션 진행하면서 참고 부탁드리겠습니다!
다음 단계도 화이팅이에요 👏
